Create extensions_v2.md (#8495)
dotnet/csharplang: proposals/extensions_v2.md

# Modern Extensions

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

Many thanks to those who helped with this proposal.  Esp. @jnm2!

## Summary
[summary]: #summary

> Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods and constructors, in a clean and cohesive fashion.

Reviewer: Will this include operators? While I think we should push to initially preview this without operators, the final version of extensions likly needs to include them.

Author: Yes.  THis includes operators.

This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in semantically *identical* fashion (both at a source and ABI level).

A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:

> Note: the syntax is intentionally a strawman to aid in discussion.  It is not intended to be final form.

Author: ```suggestion
Note: the syntax is intentionally a strawman to aid in discussion.  It is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.
```

```c#
extension E
{

>     // Method form, replaces `public static int M<X>(SomeType<X> t, ...) { } 

Reviewer 2: ```suggestion
    // Method form, replaces `public static int M<X>(this SomeType<X> t, ...) { } 
```

```c#
// Existing style
static class E
{
    static TField field;
    static int Property => ...
    static void NonExtensionHelperMethod() { }

    static int ExtensionMethod(this string x, ...) { }
    static T GenericExtensionMethod<T, U>(this U u, ...) { }
}

// New style
extension E
{
    // Non extensions stay exactly the same.
    static TField field;
    static int Property => ...

    // Note the lack of a 'for-clause'.  This is a normal static method.
    // An *modern static extension method* will have a 'for-clause' on it
    static void NonExtensionHelperMethod() { }

    // Migrated *instance* extension members
    int ExtensionMethod(...) for string x { }
    T GenericExtensionMethod<T, U>(...) for U u { }
}
```

In other words, all existing extension methods drop `static` from their signature, and move their first parameter to a `for-clause` placed within the method header (currently strawmanned as after the parameter list).  Note: the syntax of a `for-clause` is `for parameter`, allowing things like a parameter name to be specified.  `parameter` is critical in this design to ensure the classic extension method `this` parameter can always cleanly move.

> This location cleanly supports clauses, being already where the type parameter constraint clauses go.  The extension itself (E) will get emitted exactly as a static class would be that contains extension methods (allowing usage from older compilers and other languages without any updates post this mechanical translation).

Author: ```suggestion
The strawman chooses this location as it already cleanly supports clauses, being where the type parameter constraint clauses already go.  The extension itself (E) will get emitted exactly as a static class would be that contains extension methods (allowing usage from older compilers and other languages without any updates post this mechanical translation).
```

> In other words, all existing extension methods drop `static` from their signature, and move their first parameter to a `for-clause` placed within the method header (currently strawmanned as after the parameter list).  Note: the syntax of a `for-clause` is `for parameter`, allowing things like a parameter name to be specified.  `parameter` is critical in this design to ensure the classic extension method `this` parameter can always cleanly move.
> 
> The strawman chooses this location as it already cleanly supports clauses, being where the type parameter constraint clauses already go.  The extension itself (E) will get emitted exactly as a static class would be that contains extension methods (allowing usage from older compilers and other languages without any updates post this mechanical translation).
> 
> New extension members (beyond instance members) will need to have their metadata form decided on.  Consumption from older compilers and different languages of these new members will be specified at a later point in time.

Author: ```suggestion
In other words, all existing extension methods drop `static` from their signature, and move their first parameter to a `for-clause` placed within the method header (currently strawmanned as after the parameter list).  Note: the syntax of a `for-clause` is `for parameter`, allowing things like a parameter name to be specified.  `parameter` is critical in this design to ensure the classic extension method `this` parameter can always cleanly move.

The strawman chooses this location as it already cleanly supports clauses, being where the type parameter constraint clauses already go. 

The extension itself (E) will get emitted exactly as a static class would be that contains extension methods (allowing usage from older compilers and other languages without any updates post this mechanical translation).

New extension members (beyond instance members) will need to have their metadata form decided on.  Consumption from older compilers and different languages of these new members will be specified at a later point in time.
```

```c#
extension E
{
    // All extension members extend the same thing:

    public void M() for SomeType { ... }
    public int P for SomeType { get { ... } }
    public static operator+(...) for SomeType { ... }
    // etc
}

// Can be translated to:

extension E for SomeType
{
    public void M() { ... }
    public int P { get { ... } }
    public static operator+(...) { ... }
}
```

TODO: Do an ecosystem check on what percentage of existing extensions could use this simpler form.

TODO: It's possible someone might have an extension where almost all extensions extend a single type, and a small handful do something slightly different (perhaps extending by `this ref`).  Would it be beneficial here to *still* allow the extension members to provide a `for-clause` to override that default for that specific member.  For example:

```c#
extension StringExtensions for string
{
    // Lots of normal extension methods on string ...

    // Override here to extend `string?`
    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
    {
    }
}
```

It seems like this would be nice to support with little drawback.

## Expansion 2: Optional syntactic components

> As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:

Reviewer: Can we explicitly list out all of the patterns we want to support. The ones that come to mind for me are:

```csharp
for ref Span<T>
for ref readonly Span<T>
for in Span<T>
for scoped ref Span<T>
for scoped Span<T>
```

Think that is the exhaustive list.

Author: 1.  Yes.  Happy to call that out.
2. Yes.  This proposal supports all those forms.

Author: i have called all thse forms out explicit above now.

Reviewer 2: Is `scoped ref readonly` an interesting form here?

    {
        if (value is null)
            value = 0;
    }
}
```

For this reason, the strawman syntax for this `for clause` is `for parameter`, where `parameter` is the familiar:

```g4
parameter
    | attributes? modifiers? type identifier
    ;
```

(fortunately, extension methods today don't support a default value for the `this` parameter, so wel don't have to support that).

However, for many extensions no name is really required.  All extension members (except for extension-constructors and extension-static-methods) are conceptually a way to extend `this` with new functionality.  This is so much so the case that we even designed classic extension methods to use the `this` keyword as their designator.  As such, we forsee potentially making the name optional, allowing one to write an extension like so:

```c#
extension Enumerable
{
    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
        for IEnumerable<TSource> // no name
        where TResult : struct, INumber<TResult>
        where TAccumulator : struct, INumber<TAccumulator>
    {
        // Use 'this' in here to represent the value being extended
    }
}
```

> This would have to come with some default name chosen by the language for the parameter in metadata.  But that never be needed by anyone calling it from a modern compiler.

Reviewer: The proposal is aimed at compat and to hit that I do think we need to solve the parameter name of the `this`. There will inevitably be some cusotmer that has bet on calling the method in `static` form using named arguments. If we don't have a solution for that then I fear that teams like runtime will have hesitation in migrating to the new syntax. 

I do wonder if we would want to split that out though into a separate proposal. There are a number of other cases where we've desired to allow named arguments to use a different name than the parameter name within the type:

1. For records there is the tension of do you use pascal or camel case for parameter names? For .NET style you want the field to be pascal cased but the parameter  camel cased. 
2. For dictionary expressions one of the comparer designs includes binding to a parameter named `comparer`. Unfortunately there are plenty of examples where that name was not used for the comparer. Have to decide between compromising the design or breaking compat.

Maybe this design is the final straw in saying lets go and solve that problem generally so we can unlock a number of scenarios. Essentially give developers a way of saying 

```csharp
[ParameterName(0, "stringComparer")]
public void M(IComparer<string> comparer) { ... }

M(stringComparer: null); // okay
M(comparer: null); // okay
```

Author: > The proposal is aimed at compat and to hit that I do think we need to solve the parameter name of the this. There will inevitably be some cusotmer that has bet on calling the method in static form using named arguments. If we don't have a solution for that then I fear that teams like runtime will have hesitation in migrating to the new syntax.

The proposal allows the name to be specified just as you can todaay.  It's only in the "future enhancements" where we even consider allowing that to be optional.  

My goal is that hte *core* proposal give *perfect* binary compat.  in other words it generates *identical* IL/metadata.

Author: > Maybe this design is the final straw in saying lets go and solve that problem generally so we can unlock a number of scenarios. Essentially give developers a way of saying

My preference is that that is a future enhancement.  I think i want to start us with perfect compat, and then add niceties for cases where it either isn't needed, or can be made simpler.

Reviewer: I'm confused how this design achieves perfect compat then. Consider the following: 

```csharp
1.M();
E.M(i: 42);

static class E { 
  internal static void M(this int i) { }
}
```

How can I port this to the new model without the ability to name the `this`?

Author: you can name the `this`.  That's exactly what the spec says (and even gives examples).  The for-clause is `for parameter`.  so you write this as:

```c#
extension E
{
    internal void M() for int i { /*exact same body*/ }
}
```

Reviewer: Ah okay. That works for this then.

Author: Here's the section on how to migrate:

> In other words, all existing extension methods drop `static` from their signature, and move their first parameter to a `for-clause` placed within the method header (currently strawmanned as after the parameter list).  Note: the syntax of a `for-clause` is `'for' parameter`, allowing things like a parameter name to be specified.  `parameter` is critical in this design to ensure the classic extension method `this` parameter can always cleanly move.

> The strawman chooses this location as it already cleanly supports clauses, being where the type parameter constraint clauses already go. 

We can write up a migration guide (and we'll have a fixer).  but the steps (i believe) are simply:

1. change `static class` to `extension`.  Keep the type header otherwise the same.
2. for all extension methods in the static class (those that use `this` in the first parameter):
  a. drop 'static'
  c. remove the `this` parameter from the parameter list.
  d. make a `for-clause` with the original `this` parameter (with the `this` keyword dropped)
  e. place the `for-clause` after the existing parameter list, before any constraints

I believe that's the entire algorithm.  

Conversely, when compiling for *instance extension methods*, we will do *everything* (binding, reducing, lowering, emitting) as if it was declared in the original static-extension-method form where the for-clause is the first parameter.  I've tried very hard to make all sides of this very simple, so the new instance extension methods and the classic static extensions are *trivially* convertible.  We need this (imo) to help ensure the perfect source and binary compat.

This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in semantically *identical* fashion (both at a source and ABI level).

A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:

Note: the syntax is intentionally a strawman to aid in discussion.  It is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.

```c#
extension E
{
    // Method form, replaces `public static int M<X>(this SomeType<X> t, ...) { } 
    public int M<X>(...) for SomeType<X> t { }

    // Property form:
    public int Count<X> for SomeType<X> { get { ... } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> { get { ... } }

    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

>     // Constructor form:
>     public SomeType<X>() for SomeType<X>
> 
>     // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
>     public static bool Boo(string s1, string s2) for Assert { }
> 

Author: ```suggestion
    // Constructor form:
    public SomeType<X>() for SomeType<X>

    // Event form:
    // Note: would have to be the add/remove form.  field-backed events would not be possible.
    public event Action E<X> for SomeType<X> { add { } remove { } }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }
```

}
```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

> // For method/property/indexer/operator/constructor declarations

Author: ```suggestion
// For method/property/indexer/operator/constructor/event declarations
```

// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration
    | attributes? modifiers identifier type-parameters for-clause property-body;
```

> Modern extensions continue to not allow adding fields or destructors to a type.

Author: ```suggestion
Modern extensions continue to not allow adding fields or destructors to a type.
```

    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

    // Constructor form:
    public SomeType<X>() for SomeType<X>

    // Event form:
    // Note: would have to be the add/remove form.  field-backed events would not be possible.
    public event Action E<X> for SomeType<X> { add { } remove { } }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }
}
```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

// For method/property/indexer/operator/constructor/event declarations
// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration
    | attributes? modifiers identifier type-parameters for-clause property-body;
```

> Modern extensions continue to not allow adding fields or destructors to a type.

Author: ```suggestion
Modern extensions continue to not allow adding fields or destructors to a type.

Open question: Support nested types? This seems reasonable and the same as adding any other static extension member.
```

# Modern Extensions

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

Many thanks to those who helped with this proposal.  Esp. @jnm2!

## Summary
[summary]: #summary

> Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods and constructors, in a clean and cohesive fashion.

Author: ```suggestion
Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.
```

> This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in semantically *identical* fashion (both at a source and ABI level).

Author: ```suggestion
This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).
```

A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:

> Note: the syntax is intentionally a strawman to aid in discussion.  It is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.

Author: ```suggestion
Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.
```

```c#
extension E
{

>     // Method form, replaces `public static int M<X>(this SomeType<X> t, ...) { } 

Author: ```suggestion
    // Instance method form, replaces `public static int M<X>(this SomeType<X> t, ...) { } 
```

    public int M<X>(...) for SomeType<X> t { }

>     // Property form:
>     public int Count<X> for SomeType<X> { get { ... } }
> 
>     // Indexer form:
>     public int this<T>[int index] for SomeType<X> { get { ... } }
> 
>     // Operator form:
>     public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
> 
>     // Constructor form:
>     public SomeType<X>() for SomeType<X>
> 
>     // Event form:
>     // Note: would have to be the add/remove form.  field-backed events would not be possible.
>     public event Action E<X> for SomeType<X> { add { } remove { } }

Author: ```suggestion
    // Property form:
    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
    public int Count<X> for SomeType<X> { get { ... } }

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.
    public event Action E<X> for SomeType<X> { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> { get { ... } }

    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

    // Constructor form:
    public SomeType<X>() for SomeType<X>
```

}
```

## Disambiguation

Classic extension methods today can be disambiguated by falling back to static-invocation syntax.  For example, if `x.Count()` is ambiguous, it is possible to switch to some form of `StaticClass.Count(x)` to call the desired method.  A similar facility is needed for modern extension methods.  While the existing method-invocation-translation approach works fine for methods (where the receiver can be remapped to the first argument of the static extension method call), it is ungainly for these other extension forms.

As an initial strawman this proposal suggests reusing `cast expression` syntax for disambiguation purposes.  For example:

```c#
var v1 = ((Extension)receiver).ExtensionMethod(); // instead of Extension.ExtensionMethod(receiver)
var v2 = ((Extension)receiver).ExtensionProperty;
var v3 = ((Extension)receiver)[indexerArg];
var v4 = (Extension)receiver1 + receiver2;
```

Constructors and static methods would not need any special syntax as the extension can cleanly be referenced as a type where needed.

```c#
var v3 = new Extension(...); // Makes instance of the actual extended type.
var v4 = Extension.StaticExtensionMethod(...);
```

> Note 1: while the cast syntax traditionally casts or converts a value, that would not be the case for its use here.  It would only be used as a lookup mechanism to indicate which extension gets priority.  Importantly, even with this syntax, extensions themselves are not types.  For example:

Reviewer 2: Is `typeof(Extension)` also blocked? `nameof(Extension)`?

Author: great points.  no.  neither woudl be blocked.  will call out.

```c#
Extension e1;                   // Not legal.  Extension is not a type.
Extension[] e2;                 // Not legal.  Extension is not a type.
List<Extension> e3;             // Not legal.  Extension is not a type.
var v1 = (Extension)receiver;   // Not legal.  Can't can't have a value of extension type.
```

This is exactly the same as the restrictions on static-types *except* with the carve out that you can use the extension in a cast-syntax or new-expression *only* for lookup purposes and nothing else.

Note 2. If cast syntax is not desirable here (especially if confuses the idea if extensions are types), we can come up with a new syntactic form.  We are not beholden to the above syntax.

# Future expansion

The above initial strawman solves several major goals for we want for the extensions space:

1. Supporting a much broader set of extension member types.
2. Having a clean syntax for extension members that matches the regular syntax form (in other words, an extension proeprty still looks like a property).

> 3. Ensuring teams can move safely to modern extensions *especially* in environments where source binary compatibility is non-negotiable.

Author: ```suggestion
3. Ensuring teams can move safely to modern extensions *especially* in environments where source *and* binary compatibility is non-negotiable.
```

> However, there are parts of its core design that are not ideal which we would like to ensure we can expand on.  These expansions could be released with extensions if time and other resources permit.  Or they could came later and cleanly sit on top of the feature to improve the experience.

Author: ```suggestion
However, there are parts of its core design that are not ideal in the long term which we would like to ensure we can expand on.  These expansions could be released with extensions if time and other resources permit.  Or they could came later and cleanly sit on top of the feature to improve the experience.
```

These areas are:

## Expansion 1: Syntactic clumsiness and repetition

> The initial extension form values source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:

Reviewer 2: Garden-path sentence, `values` sounds like a noun for a while past reading it
```suggestion
The initial extension form places value on source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:
```

Reviewer 2: Or:
```suggestion
In this initial form for extensions, value is placed on source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:
```

1. bifurcation; where some codebases adopt modern extensions and some do not.
2. internal inconsistency; where some codebases must keep around old extensions and new extensions, with confusion about the semantics of how each interacts with the other.

> Because classic extension methods have very few restrictions, modern extension methods need to be flexible enough to support all the scenarios supported there.

Reviewer 2: ```suggestion
Because classic extension methods have very few restrictions, modern extension methods need to be flexible enough to support all the scenarios which they support.
```

> However, many codebases do not need all the flexibility that classic extension methods afforded.  For example, classic extension methods allow disparate extension methods in a single static class to target multiple different types.  For use cases where that isn't required, we forsee a natural extension (pun intended) where one can modern translate extensions like so:

Author: ```suggestion
However, many codebases do not need all the flexibility that classic extension methods afforded.  For example, classic extension methods allow disparate extension methods in a single static class to target multiple different types.  For use cases where that isn't required, we forsee a natural extension (pun intended) where one can translate a modern extension like so:
```

    public event Action E<X> for SomeType<X> { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> { get { ... } }

    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

    // Constructor form:
    public SomeType<X>() for SomeType<X>

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }
}
```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

// For method/property/indexer/operator/constructor/event declarations
// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration
    | attributes? modifiers identifier type-parameters for-clause property-body;

> ```

Author: ```suggestion
```

The use of `parameter` means all of the following are legal, with the same semantic extension methods have today:

```c#
for ref Span<T>
for ref readonly Span<T>
for in Span<T>
for scoped ref Span<T>
for scoped Span<T>
```
```

Modern extensions continue to not allow adding fields or destructors to a type.

Open question: Support nested types? This seems reasonable and the same as adding any other static extension member.

## Migration and compatibility

> Given an existing static class with extensions, a straightforward *semantically identical* translation to modern extensions is done in the following fashion.

Author: ```suggestion
Given an existing static class with extensions, a straightforward *semantically identical* (both at the source and binary level) translation to modern extensions is done in the following fashion.
```

# Modern Extensions

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

Many thanks to those who helped with this proposal.  Esp. @jnm2!

## Summary
[summary]: #summary

Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.

> This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).

Author: ```suggestion
This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).

Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with existing extensions, and then potential enhancements on top of this to make certain common designs more succinct and pleasant.
```

```c#
var v1 = ((Extension)receiver).ExtensionMethod(); // instead of Extension.ExtensionMethod(receiver)
var v2 = ((Extension)receiver).ExtensionProperty;
var v3 = ((Extension)receiver)[indexerArg];
var v4 = (Extension)receiver1 + receiver2;
```

Constructors and static methods would not need any special syntax as the extension can cleanly be referenced as a type where needed.

```c#
var v3 = new Extension(...); // Makes instance of the actual extended type.
var v4 = Extension.StaticExtensionMethod(...);
```

Note 1: while the cast syntax traditionally casts or converts a value, that would not be the case for its use here.  It would only be used as a lookup mechanism to indicate which extension gets priority.  Importantly, even with this syntax, extensions themselves are not types.  For example:

```c#
Extension e1;                   // Not legal.  Extension is not a type.
Extension[] e2;                 // Not legal.  Extension is not a type.
List<Extension> e3;             // Not legal.  Extension is not a type.
var v1 = (Extension)receiver;   // Not legal.  Can't can't have a value of extension type.
```

> This is exactly the same as the restrictions on static-types *except* with the carve out that you can use the extension in a cast-syntax or new-expression *only* for lookup purposes and nothing else.

Author: ```suggestion
This is exactly the same as the restrictions on static-types *except* with the carve out that you can use the extension in a cast-syntax or new-expression *only* for lookup purposes, or where a static-class could be used, and nothing else.
```

> This is exactly the same as the restrictions on static-types *except* with the carve out that you can use the extension in a cast-syntax or new-expression *only* for lookup purposes, or where a static-class could be used, and nothing else.  Usage in places like `nameof(Extension)` or `typeof(Extension)` would still be fine, as those are places where a static type is allowed.

Author: @jnm2 help me wordsmith :D

Reviewer 2: Looks fine to me!

I could imagine confusion that `Activator.CreateInstance(typeof(Extension))` and `new Extension()` mean different things, so I'm weighing the use cases of even having `typeof` and `nameof`.

Author: we need typeof/nameof for perfect compat.

Author: any place a static-class can be used, you must be able to use an extensino.  or else we break compat :)

Note 2. If cast syntax is not desirable here (especially if confuses the idea if extensions are types), we can come up with a new syntactic form.  We are not beholden to the above syntax.

# Future expansion

The above initial strawman solves several major goals for we want for the extensions space:

1. Supporting a much broader set of extension member types.
2. Having a clean syntax for extension members that matches the regular syntax form (in other words, an extension proeprty still looks like a property).
3. Ensuring teams can move safely to modern extensions *especially* in environments where source *and* binary compatibility is non-negotiable.

However, there are parts of its core design that are not ideal in the long term which we would like to ensure we can expand on.  These expansions could be released with extensions if time and other resources permit.  Or they could come later and cleanly sit on top of the feature to improve the experience.

These areas are:

## Expansion 1: Syntactic clumsiness and repetition

> The initial extension form values source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:

Author: ```suggestion
The initial extension form considers source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:
```

## Summary
[summary]: #summary

Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.

This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).

Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with existing extensions, and then potential enhancements on top of this to make certain common designs more succinct and pleasant.

A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:

Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.

```c#
extension E
{

>     // Instance method form, replaces `public static int M<X>(this SomeType<X> t, ...) { } 
>     public int M<X>(...) for SomeType<X> t { }

Author: ```suggestion
    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
    public int M<X>(...) for SomeType<X> val { }
```

```c#
extension E
{
    // All extension members extend the same thing:

    public void M() for SomeType { ... }
    public int P for SomeType { get { ... } }
    public static operator+(...) for SomeType { ... }
    // etc
}

// Can be translated to:

extension E for SomeType
{
    public void M() { ... }
    public int P { get { ... } }
    public static operator+(...) { ... }
}
```

TODO: Do an ecosystem check on what percentage of existing extensions could use this simpler form.

TODO: It's possible someone might have an extension where almost all extensions extend a single type, and a small handful do something slightly different (perhaps extending by `this ref`).  Would it be beneficial here to *still* allow the extension members to provide a `for-clause` to override that default for that specific member.  For example:

```c#

> extension StringExtensions for string

Reviewer 2: For clarity, since the concept of dropping the identifier hasn't yet been introduced:
```suggestion
extension StringExtensions for string str
```

{
    // Lots of normal extension methods on string ...

    // Override here to extend `string?`
    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
    {
    }
}
```

It seems like this would be nice to support with little drawback.

## Expansion 2: Optional syntactic components

As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:

```c#
static class Extensions
{
    // Yes, this is legal
    public static void MakeNonNull([Attr] this ref int? value)
    {
        if (value is null)
            value = 0;
    }
}
```

For this reason, the strawman syntax for this `for clause` is `for parameter`, where `parameter` is the familiar:

```g4
parameter

>     | attributes? modifiers? type identifier

Author: ```suggestion
    | attributes? modifiers? type identifier
    | attributes? modifiers? type identifier = expression
```

    ;
```

> (fortunately, extension methods today don't support a default value for the `this` parameter, so wel don't have to support that).

Reviewer 2: ```suggestion
(Fortunately, extension methods today don't support a default value for the `this` parameter, so we don't have to support that.)
```

> Fortunately, extension methods today don't support a default value for the `this` parameter, so we don't have to support that.

Author: ```suggestion
Fortunately, extension methods today don't support a default value for the `this` parameter, so we don't have to support migrating the second `= value` form forward, and we would consider writing a default value in a `for-clause` to be an error.
```

> (Fortunately, extension methods today don't support a default value for the `this` parameter, so we don't have to support that.)

Author: ```suggestion
Fortunately, extension methods today don't support a default value for the `this` parameter, so we don't have to support that.
```

# Modern Extensions

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

Many thanks to those who helped with this proposal.  Esp. @jnm2!

## Summary
[summary]: #summary

Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.

This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).

> Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with existing extensions, and then potential enhancements on top of this to make certain common designs more succinct and pleasant.

Author: ```suggestion
Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with classic extension methods, and then potential enhancements on top of this to make certain common designs more succinct and pleasant.  This does not preclude the feature shipping with any or all of these enhancements at launch.  It simply allows the design to be broken into separable concerns, with a clear dependency order.
```

> Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with classic extension methods, and then potential enhancements on top of this to make certain common designs more succinct and pleasant.  This does not preclude the feature shipping with any or all of these enhancements at launch.  It simply allows the design to be broken into separable concerns, with a clearer ordering of dependencies.

Author: ```suggestion
Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with classic extension methods, and then potential enhancements on top of this kernel to make certain common designs more succinct and pleasant.  This does not preclude the feature shipping with any or all of these enhancements at launch.  It simply allows the design to be broken into separable concerns, with a clearer ordering of dependencies.
```

TODO: Do an ecosystem check on what percentage of existing extensions could use this simpler form.

TODO: It's possible someone might have an extension where almost all extensions extend a single type, and a small handful do something slightly different (perhaps extending by `this ref`).  Would it be beneficial here to *still* allow the extension members to provide a `for-clause` to override that default for that specific member.  For example:

```c#
extension StringExtensions for string str
{
    // Lots of normal extension methods on string ...

    // Override here to extend `string?`
    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
    {
    }
}
```

It seems like this would be nice to support with little drawback.

## Expansion 2: Optional syntactic components

As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:

```c#
static class Extensions
{
    // Yes, this is legal
    public static void MakeNonNull([Attr] this ref int? value)
    {
        if (value is null)
            value = 0;
    }
}
```

> For this reason, the strawman syntax for this `for clause` is `for parameter`, where `parameter` is the familiar:

Author: ```suggestion
For this reason, the strawman syntax for this `for-clause` is `'for' parameter`, where `parameter` is the familiar:
```

> For this reason, the strawman syntax for this `for-clause` is `'for' parameter`, where `parameter` is the familiar:
> 
> ```g4
> parameter
>     | attributes? modifiers? type identifier
>     | attributes? modifiers? type identifier '=' expression
>     ;
> ```

Author: ````suggestion
For this reason, the strawman syntax is:

```g4
for-clause
    | 'for' parameter`
    ;

parameter (unchanged)
    | attributes? modifiers? type identifier
    | attributes? modifiers? type identifier '=' expression
    ;
```
````

```c#
extension E
{
    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
    public int M<X>(...) for SomeType<X> val { }

    // Property form:
    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
    public int Count<X> for SomeType<X> val { get { ... } }

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.
    public event Action E<X> for SomeType<X> val { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> val { get { ... } }

    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

    // Constructor form:
    public SomeType<X>() for SomeType<X> { }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }

> }

Author: ```suggestion

    // Static extensions of properties, events supported.  They are supported in static classes
    // Nested types supported.  They are supported in static classes.
}
```

{
    // Lots of normal extension methods on string ...

    // Override here to extend `string?`
    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
    {
    }
}
```

It seems like this would be nice to support with little drawback.

## Expansion 2: Optional syntactic components

As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:

```c#
static class Extensions
{
    // Yes, this is legal
    public static void MakeNonNull([Attr] this ref int? value)
    {
        if (value is null)
            value = 0;
    }
}
```

For this reason, the strawman syntax is:

```g4
for-clause
    | 'for' parameter`
    ;

parameter (unchanged)
    | attributes? modifiers? type identifier
    | attributes? modifiers? type identifier '=' expression
    ;
```

Fortunately, extension methods today don't support a default value for the `this` parameter, so we don't have to support migrating the second `= value` form forward, and we would consider writing a default value in a `for-clause` to be an error.

> However, for many extensions no name is really required.  All instance extension members (methods, properties, indexers and events) are conceptually a way to extend `this` with new functionality.  This is so much so the case that we even designed classic extension methods to use the `this` keyword as their designator.  As such, we forsee potentially making the name optional, allowing one to write an extension like so:

Author: ```suggestion
However, for many extensions no name is really required.  All non-static extension members (instance methods, properties, indexers and events) are conceptually a way to extend `this` with new functionality.  This is so much so the case that we even designed classic extension methods to use the `this` keyword as their designator.  As such, we forsee potentially making the name optional, allowing one to write an extension like so:
```

> However, for many extensions no name is really required.  All extension members (except for extension-constructors and extension-static-methods) are conceptually a way to extend `this` with new functionality.  This is so much so the case that we even designed classic extension methods to use the `this` keyword as their designator.  As such, we forsee potentially making the name optional, allowing one to write an extension like so:

Author: ```suggestion
However, for many extensions no name is really required.  All instance extension members (methods, properties, indexers and events) are conceptually a way to extend `this` with new functionality.  This is so much so the case that we even designed classic extension methods to use the `this` keyword as their designator.  As such, we forsee potentially making the name optional, allowing one to write an extension like so:
```

* [ ] Specification: Not Started

Many thanks to those who helped with this proposal.  Esp. @jnm2!

## Summary
[summary]: #summary

Modern Extensions introduce a new syntax to produce "extension members", greatly expanding on the set of supported members including properties, static methods, operators and constructors (and more), in a clean and cohesive fashion.

This new form subsumes C# 3's "classic extension methods", allowing migration to the new modern form in a semantically *identical* fashion (both at a source and ABI level).

Note: this proposal is broken into two parts.  A core kernel needed to initially ship with perfect forward and backwards compatibility with classic extension methods, and then potential enhancements on top of this kernel to make certain common designs more succinct and pleasant.  This does not preclude the feature shipping with any or all of these enhancements at launch.  It simply allows the design to be broken into separable concerns, with a clearer ordering of dependencies.

A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:

> Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  A form that is only good for properties, but not for other members, it likely not appropriate.

Author: ```suggestion
Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  For example, a form that is only good for properties, but not for other members, it likely not appropriate.
```

```c#
extension E
{
    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
    public int M<X>(...) for SomeType<X> val { }

    // Property form:
    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.

>     public int Count<X> for SomeType<X> { get { ... } }

Reviewer 2: It's confusing me that many of these don't have a parameter name, since the core proposal isn't making the parameter name optional on instance extensions.

Author: ok.  adding.

>     public int Count<X> for SomeType<X> { get { ... } }

Author: ```suggestion
    public int Count<X> for SomeType<X> val { get { ... } }
```

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.

>     public event Action E<X> for SomeType<X> { add { } remove { } }

Author: ```suggestion
    public event Action E<X> for SomeType<X> val { add { } remove { } }
```

>     public event Action E<X> for SomeType<X> val { add { } remove { } }

Reviewer 2: Can we allow _static_ extension properties and events to be field-backed?

Author: yes.

    // Indexer form:

>     public int this<T>[int index] for SomeType<X> { get { ... } }

Author: ```suggestion
    public int this<T>[int index] for SomeType<X> val { get { ... } }
```

    // Operator form:

>     public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

Author: note: no `SomeType<X> val`, an operator is static, so it is not passed an instance 'this'.

Reviewer 2: That could be a great note to put in the comment lines.

>     public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

Reviewer 2: This space has always been present (or is this intentional?)
```suggestion
    public static SomeTypeX<X> operator + <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
```

Author: keeping like this for now.

>     public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

Reviewer 2: Can I see a conversion operator form too?

    // Constructor form:

>     public SomeType<X>() for SomeType<X>

Author: ```suggestion
    public SomeType<X>() for SomeType<X> { }
```
note: no `SomeType<X> val`, an constructor is static, so it is not passed an instance 'this'.

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`

>     public static bool Boo(string s1, string s2) for Assert { }

Author: note: no `SomeType<X> val`, an static method is static, so it is not passed an instance 'this'.

}
```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

// For method/property/indexer/operator/constructor/event declarations
// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration
    | attributes? modifiers identifier type-parameters for-clause property-body;
```

The use of `parameter` means all of the following are legal, with the same semantic extension methods have today:

```c#

> for ref Span<T>
> for ref readonly Span<T>
> for in Span<T>
> for scoped ref Span<T>
> for scoped Span<T>

Author: ```suggestion
for ref Span<T> span
for ref readonly Span<T> span
for in Span<T> span
for scoped ref Span<T> span
for scoped Span<T> span
```

```

Modern extensions continue to not allow adding fields or destructors to a type.

Open question: Support nested types? This seems reasonable and the same as adding any other static extension member.

## Migration and compatibility

Given an existing static class with extensions, a straightforward *semantically identical* (both at the source and binary level) translation to modern extensions is done in the following fashion.

```c#
// Existing style
static class E
{
    static TField field;
    static int Property => ...
    static void NonExtensionHelperMethod() { }

    static int ExtensionMethod(this string x, ...) { }
    static T GenericExtensionMethod<T, U>(this U u, ...) { }
}

// New style
extension E
{
    // Non extensions stay exactly the same.
    static TField field;
    static int Property => ...

    // Note the lack of a 'for-clause'.  This is a normal static method.
    // An *modern static extension method* will have a 'for-clause' on it
    static void NonExtensionHelperMethod() { }

    // Migrated *instance* extension members
    int ExtensionMethod(...) for string x { }
    T GenericExtensionMethod<T, U>(...) for U u { }
}
```

> In other words, all existing extension methods drop `static` from their signature, and move their first parameter to a `for-clause` placed within the method header (currently strawmanned as after the parameter list).  Note: the syntax of a `for-clause` is `'for' parameter`, allowing things like a parameter name to be specified.  `parameter` is critical in this design to ensure the classic extension method `this` parameter can always cleanly move.
> 
> The strawman chooses this location as it already cleanly supports clauses, being where the type parameter constraint clauses already go. 

Author: ```suggestion
In other words, all existing extension methods drop `static` from their signature, and move their first parameter to a `for-clause` placed within the method header (currently strawmanned as after the parameter list).  The strawman chooses this location as it already cleanly supports clauses, being where the type parameter constraint clauses already go.

Note: the syntax of a `for-clause` is `'for' parameter`, allowing things like a parameter name to be specified.  `parameter` is critical in this design to ensure the classic extension method `this` parameter can always cleanly move. 
```

A full example of this translation with a real world complex signature would be:

```c#
static class Enumerable
{
    public static TResult Sum<TSource, TResult, TAccumulator>(this IEnumerable<TSource> source, Func<TSource, TResult> selector)
        where TResult : struct, INumber<TResult>
        where TAccumulator : struct, INumber<TAccumulator>
    {
        // Original body
    }
}

extension Enumerable
{
    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
        for IEnumerable<TSource> source
        where TResult : struct, INumber<TResult>
        where TAccumulator : struct, INumber<TAccumulator>
    {
        // Exactly the same code as original body.
    }
}
```

This form supports *non* extension static methods.  For example `Enumerable.Range` would migrate like so:

```c#
static class Enumerable
{
    public static IEnumerable<int> Range(int start, int count) { ... }
}

extension Enumerable
{
    // Exact same signature.  No 'for-clause'.
    public static IEnumerable<int> Range(int start, int count) { ... }
}
```

## Disambiguation

> Classic extension methods today can be disambiguated by falling back to static-invocation syntax.  For example, if `x.Count()` is ambiguous, it is possible to switch to some form of `StaticClass.Count(x)` to call the desired method.  A similar facility is needed for modern extension methods.  While the existing method-invocation-translation approach works fine for methods (where the receiver can be remapped to the first argument of the static extension method call), it is ungainly for these other extension forms.

Author: ```suggestion
Classic extension methods today can be disambiguated by falling back to static-invocation syntax.  For example, if `x.Count()` is ambiguous, it is possible to switch to some form of `StaticClass.Count(x)` to call the desired method.  A similar facility is needed for modern extension members.  While the existing method-invocation-translation approach works fine for methods (where the receiver can be remapped to the first argument of the static extension method call), it is ungainly for these other extension forms.
```

These areas are:

## Expansion 1: Syntactic clumsiness and repetition

The initial extension form considers source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:
1. bifurcation; where some codebases adopt modern extensions and some do not.
2. internal inconsistency; where some codebases must keep around old extensions and new extensions, with confusion about the semantics of how each interacts with the other.

Because classic extension methods have very few restrictions, modern extension methods need to be flexible enough to support all the scenarios which they support.

However, many codebases do not need all the flexibility that classic extension methods afforded.  For example, classic extension methods allow disparate extension methods in a single static class to target multiple different types.  For use cases where that isn't required, we forsee a natural extension (pun intended) where one can translate a modern extension like so:

> ```c#
> extension E
> {
>     // All extension members extend the same thing:
> 
>     public void M() for SomeType { ... }
>     public int P for SomeType { get { ... } }
>     public static operator+(...) for SomeType { ... }
>     // etc
> }
> 
> // Can be translated to:
> 
> extension E for SomeType
> {
>     public void M() { ... }
>     public int P { get { ... } }
>     public static operator+(...) { ... }
> }
> ```

Author: ```suggestion
```c#
extension E
{
    // All extension members extend the same thing:

    public void M() for SomeType str { ... }
    public int P for SomeType str { get { ... } }
    public static operator+(...) for SomeType str { ... }
    // etc
}

// Can be translated to:

extension E for SomeType str
{
    public void M() { ... }
    public int P { get { ... } }
    public static operator+(...) { ... }
}
```
```

```c#
extension StringExtensions for string str
{
    // Lots of normal extension methods on string ...

    // Override here to extend `string?`
    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
    {
    }
}
```

It seems like this would be nice to support with little drawback.

## Expansion 2: Optional syntactic components

As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:

```c#
static class Extensions
{
    // Yes, this is legal
    public static void MakeNonNull([Attr] this ref int? value)
    {
        if (value is null)
            value = 0;
    }
}
```

For this reason, the strawman syntax is:

```g4
for-clause
    | 'for' parameter`
    ;

parameter (unchanged)
    | attributes? modifiers? type identifier
    | attributes? modifiers? type identifier '=' expression
    ;
```

> Fortunately, extension methods today don't support a default value for the `this` parameter, so we don't have to support migrating the second `= value` form forward, and we would consider writing a default value in a `for-clause` to be an error.

Reviewer 2: Another thing `parameter` allows that `this` parameters can't use is `params` and `out`. You also wouldn't be able to use the `this` parameter modifier here :D This could be stated as, `parameter` forms that are currently allowed together with `this`, but without including `this`.

Author: sure.  please wordsmith :)

}
```

This has a few new, but solvable, design challenges.  For example, say one has the code:

```c#
List<int> ints = ...;
var v = ints.ForEach(i => Console.WriteLine(i));
```

This naturally raises the question of how does this extension get picked for this particular receiver, and how does its type parameter get instantiated to the `int` type.

Conceptually (and trying to keep somewhat in line with classic extension methods), we really want to think of the 'receiver' as an 'argument' to some method where normal type inference occurs.  Morally, we could think of there being a `IListExtension<T> Infer<T>(IList<T> list)` function whose shape is determined by the extension and its type-parameters and the extended receiver parameter.

Then, when trying to determine if an extension applies to a receiver, it would be akin to calling that function with the receiver and seeing if inference works.  In the above example that would mean performing type inference on `Infer(ints)` seeing that `T` then bound to `int`, which then gives you back `IListExtensions<int>`.  At that point, lookup would then find and perform overload resolution on `ForEach(Action<int>)` with the lambda parameter.

> This approach does fundamentally take expand on the initial extension-members approach as now calling extensions is done in two-phases.  An initial phase to determine and infer extension type parameters based on the receiver, and  a second phase to determine and perform overload resolution on the member.

Reviewer 2: ```suggestion
This approach does fundamentally expand on the initial extension-members approach, as now, calling extensions is done in two phases.  An initial phase to determine and infer extension type parameters based on the receiver, and a second phase to determine and perform overload resolution on the member.
```

We believe this is very powerful and beneficial.  But there are deep design questions here which may cause this to be scheduled after the core extension members work happens.

## Expansion 4: Extensions as actual types

> We are very undecided on if we actually want this.  Currently, our view is that it feels like 'roles' fits this goal much better, especially if roles have the ability to be 'implicit' or 'explicit'.  Extensions exist very much in the space where they are erased and really are just delicious delicious sugar over calling effectively static helpers to augment a type or value.

Reviewer 2: > especially if roles have the ability to be 'implicit' or 'explicit'.

I'm a little confused at how a role could be implicit without, unavoidably, providing extension members in a new way, unless implicit just means you can pass values without casting but lookup is unaffected. With the recent LDM decision to define roles as "explicit extensions," saying "implicit role" in this new context spins my head a bit.

> Roles, on the other hand seem more fitted to the type space where they are truly part of the type system, intending to appear in signatures, generics, and the like, potentially with strong enforcement about values moving into or out of the role.

Reviewer 2: ```suggestion
Roles, on the other hand, seem more fitted to the type space where they are truly part of the type system, intended to appear in signatures, generics, and the like, potentially with strong enforcement about values moving into or out of the role.
```

```c#
extension E
{
    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
    public int M<X>(...) for SomeType<X> val { }

    // Property form:
    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
    public int Count<X> for SomeType<X> val { get { ... } }

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.
    public event Action E<X> for SomeType<X> val { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> val { get { ... } }

    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

    // Constructor form:
    public SomeType<X>() for SomeType<X> { }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }

>     // Static extensions properties, indexers and events supported.

Author: ```suggestion
    // Static extensions properties, indexers and events are all conceptually supportable.
    // Though we can decide which are sensible to have or not.
```

```c#
extension E
{
    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
    public int M<X>(...) for SomeType<X> val { }

    // Property form:
    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
    public int Count<X> for SomeType<X> val { get { ... } }

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.
    public event Action E<X> for SomeType<X> val { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> val { get { ... } }

    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

    // Constructor form:
    public SomeType<X>() for SomeType<X> { }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }

>     // Static extensions of properties, events supported.  They are supported in static classes

Author: ```suggestion
    // Static extensions properties, indexers and events supported.
```

>     // Though we can decide which are sensible to have or not.
>     // Nested types supported.  They are supported in static classes.

Author: ```suggestion
    // Though we can decide which are sensible to have or not.
    
    // Nested types must be supported to maintain compatibility with existing static classes with extension members in them.
```

```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

// For method/property/indexer/operator/constructor/event declarations
// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration
    | attributes? modifiers identifier type-parameters for-clause property-body;
```

The use of `parameter` means all of the following are legal, with the same semantic extension methods have today:

```c#
for ref Span<T> span
for ref readonly Span<T> span
for in Span<T> span
for scoped ref Span<T> span
for scoped Span<T> span
```

Modern extensions continue to not allow adding fields or destructors to a type.

> Open question: Support nested types? This seems reasonable and the same as adding any other static extension member.

Author: ```suggestion
```

A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:

Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  For example, a form that is only good for properties, but not for other members, it likely not appropriate.

```c#
extension E
{
    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
    public int M<X>(...) for SomeType<X> val { }

    // Property form:
    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
    public int Count<X> for SomeType<X> val { get { ... } }

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.
    public event Action E<X> for SomeType<X> val { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> val { get { ... } }

>     // Operator form:
>     public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }

Author: ```suggestion
    // Operator form:
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
    
    // Conversion form:
    public static implicit operator SomeTypeX<X>(int i) for SomeType<X> { ... }
```

    // Constructor form:
    public SomeType<X>() for SomeType<X> { }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }

>     // Static extensions properties, indexers and events are all conceptually supportable.
>     // Though we can decide which are sensible to have or not.

Author: ```suggestion
    // Static extensions properties, indexers and events are all conceptually supportable.
    // Though we can decide which are sensible to have or not.
    // Static extensions can having backing static fields. Static extension properties can use `field`.
```

A full example of this translation with a real world complex signature would be:

```c#
static class Enumerable
{
    public static TResult Sum<TSource, TResult, TAccumulator>(this IEnumerable<TSource> source, Func<TSource, TResult> selector)
        where TResult : struct, INumber<TResult>
        where TAccumulator : struct, INumber<TAccumulator>
    {
        // Original body
    }
}

extension Enumerable
{
    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
        for IEnumerable<TSource> source
        where TResult : struct, INumber<TResult>
        where TAccumulator : struct, INumber<TAccumulator>
    {
        // Exactly the same code as original body.
    }
}
```

This form supports *non* extension static methods.  For example `Enumerable.Range` would migrate like so:

```c#
static class Enumerable
{
    public static IEnumerable<int> Range(int start, int count) { ... }
}

extension Enumerable
{
    // Exact same signature.  No 'for-clause'.
    public static IEnumerable<int> Range(int start, int count) { ... }
}
```

> ## Disambiguation

Author: ```suggestion
Perfect source and binary compatibility are goals here.  That means any other features that work with static classes and extensions are expected to migrate over to extensions without change.  This includes, but it not limited to other features like 'attributes'.  For example, all attributes that might be present on a `static class` should migrate unchanged to an `extension`.  Other features not mentioned here should not be inferred to not be part of this design.  By default all features should continue being compatible, and only explicitly specified deviations should be allowed.

## Disambiguation
```

    // Constructor form:
    public SomeType<X>() for SomeType<X> { }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    public static bool Boo(string s1, string s2) for Assert { }

    // Static extensions properties, indexers and events are all conceptually supportable.
    // Though we can decide which are sensible to have or not.
    // Static extensions can having backing static fields. Static extension properties can use `field`.
    
    // Nested types must be supported to maintain compatibility with existing static classes with extension members in them.
}
```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

// For method/property/indexer/operator/constructor/event declarations
// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration

>     | attributes? modifiers identifier type-parameters for-clause property-body;

Author: ```suggestion
    | attributes? modifiers identifier type-parameters for-clause property-body
    ;
    
compilation-unit-member
    | ...
    | extension
    ;
 
namespace-declaration-member
    | ...
    | extension
    ;
```

A rough strawman of the syntax is as follows.  In all cases, the extended type is shown to be generic, to indicate handling that complex case:

Note: the strawman is not intended to be final form.  That said, it is useful to see any proposed form with all members to ensure that it's generally comprehensible.  For example, a form that is only good for properties, but not for other members, it likely not appropriate.

```c#
extension E
{
    // Instance method form, replaces `public static int M<X>(this SomeType<X> val, ...) { } 
    public int M<X>(...) for SomeType<X> val { }

    // Property form:
    // Auto-properties, or usage of 'field' not allowed as extensions do not have instance state.
    public int Count<X> for SomeType<X> val { get { ... } }

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.
    public event Action E<X> for SomeType<X> val { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> val { get { ... } }

>     // Operator form:

Author: ```suggestion
    // Operator form:
    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
```

    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
    

>     // Conversion form:

Author: ```suggestion
    // Conversion form:
    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
```

    public static implicit operator SomeTypeX<X>(int i) for SomeType<X> { ... }

>     // Constructor form:

Author: ```suggestion
    // Constructor form:
    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
```

    public SomeType<X>() for SomeType<X> { }

>     // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`

Author: ```suggestion
    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    // note: no `Assert val`, a static method is not passed an instance value.
```

    // Static extensions properties, indexers and events are all conceptually supportable.
    // Though we can decide which are sensible to have or not.
    // Static extensions can having backing static fields. Static extension properties can use `field`.
    
    // Nested types must be supported to maintain compatibility with existing static classes with extension members in them.
}
```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

// For method/property/indexer/operator/constructor/event declarations
// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration
    | attributes? modifiers identifier type-parameters for-clause property-body
    ;
    
compilation-unit-member
    | ...
    | extension
    ;
 
namespace-declaration-member
    | ...
    | extension
    ;
```

> The use of `parameter` means all of the following are legal, with the same semantic extension methods have today:

Author: ```suggestion
The use of `parameter` means all of the following are legal, with the same semantics that that classic extension methods have today:
```

```c#
for ref Span<T> span
for ref readonly Span<T> span
for in Span<T> span
for scoped ref Span<T> span
for scoped Span<T> span
```

> Modern extensions continue to not allow adding fields or destructors to a type.

Author: ```suggestion
Modern extensions continue to not allow adding fields or destructors to a type.

Not all of these *new* extension member forms need be supported.  For example, we may decide that a `static extension indexer` or `static extension event` is just too esoteric, and can be cut.  The proposal shows them all though to demonstrate completeness of the idea.  All *classic* forms must be supported of course.
```

# Modern Extensions

> * [x] Proposed

Reviewer 3: The checkboxes go into issues, not into markdown proposals

    // Event form:
    // Note: would have to be the add/remove form.
    // field-backed events would not be possible as extensions do not have instance state.
    public event Action E<X> for SomeType<X> val { add { } remove { } }

    // Indexer form:
    public int this<T>[int index] for SomeType<X> val { get { ... } }

    // Operator form:
    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
    public static SomeTypeX<X> operator+ <X>(SomeType<X> s1, SomeType<X> s2) for SomeType<X> { ... }
    
    // Conversion form:
    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
    public static implicit operator SomeTypeX<X>(int i) for SomeType<X> { ... }

    // Constructor form:
    // note: no SomeType<X> val, an operator is static, so it is not passed an instance value.
    public SomeType<X>() for SomeType<X> { }

    // *Static* extension method (not possible today).  Called as `Assert.Boo("", "")`
    // note: no `Assert val`, a static method is not passed an instance value.
    public static bool Boo(string s1, string s2) for Assert { }

>     // Static extensions properties, indexers and events are all conceptually supportable.

Reviewer 3: >indexers [](http://example.com/codeflow?start=37&length=8)

I expect we'd support static extension properties, but not static extension indexers, as static indexers are not allowed (yet)

    // Though we can decide which are sensible to have or not.
    // Static extensions can having backing static fields. Static extension properties can use `field`.
    
    // Nested types must be supported to maintain compatibility with existing static classes with extension members in them.
}
```

Without specifying the full grammar changes, the intuition is that we are making the following changes:

```g4
for-clause
    | 'for' parameter
    ;

parameter (no change)
    | attributes? modifiers? type identifier ...
    ;

extension
    | attributes? modifiers? 'extension' identifier { member_declaration* }
    ;

// For method/property/indexer/operator/constructor/event declarations
// we are augmenting its syntax to allow type-parameters
// (if not already allowed) and a for-clause. For example:
property-declaration
    | attributes? modifiers identifier type-parameters for-clause property-body
    ;
    
compilation-unit-member
    | ...
    | extension
    ;
 
namespace-declaration-member
    | ...
    | extension
    ;
```

The use of `parameter` means all of the following are legal, with the same semantics that that classic extension methods have today:

```c#

> for ref Span<T> span

Reviewer 3: Personal taste, I'd put parens on anything more complicated than just-a-type, similarly to lambda parameters. Parens in `for (ref readonly Span<T> span)` help visual parsing, but they don't seem needed for `for Span<T>`.

Reviewer 3: Another example to illustrate the value of parens: `for [Attribute1][Attribute2] Type name`

Not all of these *new* extension member forms need be supported.  For example, we may decide that a `static extension indexer` or `static extension event` is just too esoteric, and can be cut.  The proposal shows them all though to demonstrate completeness of the idea.  All *classic* forms must be supported of course.

## Migration and compatibility

Given an existing static class with extensions, a straightforward *semantically identical* (both at the source and binary level) translation to modern extensions is done in the following fashion.

```c#
// Existing style
static class E
{
    static TField field;
    static int Property => ...
    static void NonExtensionHelperMethod() { }

    static int ExtensionMethod(this string x, ...) { }
    static T GenericExtensionMethod<T, U>(this U u, ...) { }
}

// New style
extension E
{
    // Non extensions stay exactly the same.
    static TField field;
    static int Property => ...

    // Note the lack of a 'for-clause'.  This is a normal static method.
    // An *modern static extension method* will have a 'for-clause' on it
    static void NonExtensionHelperMethod() { }

>     // Migrated *instance* extension members

Reviewer 3: Classic extension methods could also be allowed inside `extension`. `this` parameter on a `static` method and `for` on an instance method would be strictly equivalent.

Author: They could be. But the point of this proposal is the *minimal* core. 

It would be fine to have an augmentation to say you could still write old -styles in the new extensions.

The extension itself (E) will get emitted exactly as a static class would be that contains extension methods (allowing usage from older compilers and other languages without any updates post this mechanical translation).

New extension members (beyond instance members) will need to have their metadata form decided on.  Consumption from older compilers and different languages of these new members will be specified at a later point in time.

A full example of this translation with a real world complex signature would be:

```c#
static class Enumerable
{
    public static TResult Sum<TSource, TResult, TAccumulator>(this IEnumerable<TSource> source, Func<TSource, TResult> selector)
        where TResult : struct, INumber<TResult>
        where TAccumulator : struct, INumber<TAccumulator>
    {
        // Original body
    }
}

extension Enumerable
{
    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
        for IEnumerable<TSource> source
        where TResult : struct, INumber<TResult>
        where TAccumulator : struct, INumber<TAccumulator>
    {
        // Exactly the same code as original body.
    }
}
```

> This form supports *non* extension static methods.  For example `Enumerable.Range` would migrate like so:

Reviewer 3: >methods [](http://example.com/codeflow?start=42&length=7)

methods -> members. All non-extension static members (such as properties, fields) would work fine.

As an initial strawman this proposal suggests reusing `cast expression` syntax for disambiguation purposes.  For example:

```c#
var v1 = ((Extension)receiver).ExtensionMethod(); // instead of Extension.ExtensionMethod(receiver)
var v2 = ((Extension)receiver).ExtensionProperty;
var v3 = ((Extension)receiver)[indexerArg];
var v4 = (Extension)receiver1 + receiver2;
```

Constructors and static methods would not need any special syntax as the extension can cleanly be referenced as a type where needed.

```c#
var v3 = new Extension(...); // Makes instance of the actual extended type.
var v4 = Extension.StaticExtensionMethod(...);
```

Note 1: while the cast syntax traditionally casts or converts a value, that would not be the case for its use here.  It would only be used as a lookup mechanism to indicate which extension gets priority.  Importantly, even with this syntax, extensions themselves are not types.  For example:

```c#

> Extension e1;                   // Not legal.  Extension is not a type.

Reviewer 3: >Extension is not a type. [](http://example.com/codeflow?start=47&length=24)

nit: "Extension is not a type." That seems too strong. For static types, we just say "static types cannot be used as parameters/type arguments/..."
